src/WFQC/data.py

```diff
@@ -35,61 +36,61 @@ async def aggregate_requests(session, url):
-    """
+    try: 
+        with open(doi_library_filename, 'r') as f:
+            doi_library = json.load(f)
+    except FileNotFoundError:
+        print(f'Doi library file not found. Creating new file named {doi_library_filename}.')
+        doi_library = {} # {doi: pmid}, should I perhaps do {name: [pmid doi ]} instead?
 
+    library_updates = False
+    async with aiohttp.ClientSession() as session: 
+        for tool in tqdm(doi_tools, desc="Fetching pmids from dois."):
+            doi = tool["doi"]
 
-    # File name checking and creation 
-    date_format = "%Y%m%d"
-    pattern = f'biotools_metadata_{topicID}*'
-    matching_files = glob.glob(pattern)
-    
-    if matching_files:
-        matching_files.sort(key=os.path.getmtime)
-        csv_filename = matching_files[-1]
-        
-        # Check if file older than a week
-        file_date = datetime.strptime(csv_filename.split('_')[-1].split('.')[0], date_format)
-        if file_date < datetime.now() - timedelta(days=7):
-            print("Old datafile. Updating...")
-        else:
-            print("Bio.tools data loaded from existing CSV file.")
-            df = pd.read_csv(csv_filename)
-            return df
-    else:
-        print("No existing bio.tools CSV file. Downloading data.") 
-        # Define the CSV filename
-        csv_filename = f'biotools_metadata_{topicID}_{datetime.now().strftime(date_format)}.csv' 
-    
+            # Check if tool is already in library 
+            if doi in doi_library: 
+                doi_pmid = doi_library[doi] 
+                tool["pmid"] = doi_pmid
+                continue
+            
+            # Otherwise access NCBI API
+            library_updates = True
 
+            url = f"http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=PubMed&retmode=json&term={doi}"
+            result = await aggregate_requests(session, url)
+            try:
+                doi_pmid = str(result.get('esearchresult').get('idlist')[0])
+                if str(doi_pmid) == 'null': #this does not feel optimal 
+                    doi_pmid = None
+            except:
+                doi_pmid = None # if no pmid was found, will have to remove these nodes later
+
+            
+            tool["pmid"] = doi_pmid
 
+            if doi_pmid: # do not want to include Nones, since they might get updated in future
+                doi_library[doi] = doi_pmid
 
+    if library_updates:
+        print(f"Writing new doi, pmid pairs to file {doi_library_filename}")
+        with open(doi_library_filename, 'a') as f: #TODO a ok?
+            json.dump(doi_library, f)
     
-    # TODO: should filepath/name be allowed to be configurable?
-    # then the following could be a separate function called by this one, or is this very inefficient?
-    # TODO: should place files created in a folder named for each run
+    return doi_tools
 
-    all_tool_data = [] # TODO: predefine the length, means one more request 
+async def get_pmids(topicID):
```

Comment:
The method includes a lot of nesting, it would be good to keep that in mind, and potentially try to reduce the complexity. This often means breaking the method into multiple methods. For example, the code at line 113 after
`for tool in biotools_lst: #add tqdm here ` extracted as a method.